lib/tests.js

```diff
@@ -0,0 +1,361 @@
+                        return;
+                    }
+                    callback();
+                });
+            });
+            instance.startListening();
+            instanceTwo.startListening();
+        });
+    }
+
+
+    /**
+     * Uses multiple listeners to ensure both listeners do not get the same
+     * message.
+     *
+     * @param {Function} callback
+     */
+    function multipleListenerTest(callback) {
+        instance.on("error", (err) => {
+            callback(err);
+            instance.removeAllListeners();
+        });
+        instance.connect((err) => {
+            if (err) {
+                instance.removeAllListeners();
+                callback(err);
+
+                return;
+            }
+
+            console.log("Sending test message for multiple listener test.");
+            instance.sendMessage("multiMsg", config.topics, (testErr) => {
+                if (testErr) {
+                    instance.removeAllListeners();
+                    callback(testErr);
+
+                    return;
+                }
+
+                multipleListeners((multiErr) => {
+                    instance.removeAllListeners();
+                    instance.disconnect();
+                    callback(multiErr);
+                });
+            });
+        });
+    }
+
+
+    /**
+     * Tests wrapped messages, tests heartbeat, requeue, and remove.
+     * Tests heartbeat cleanup works as expected by asserting message is not
+     * made available again until after the heartbeat life time has elapsed.
+     *
+     * @param {Function} callback
+     */
+    function wrappedMessageTest(callback) {
+        var cleanupElapsed;
+
+        cleanupElapsed = false;
+        instance.on("error", (err) => {
+            callback(err);
+            instance.removeAllListeners();
+        });
+        instance.connect((err) => {
+            if (err) {
+                instance.removeAllListeners();
+                callback(err);
+
+                return;
+            }
+
+            console.log("Sending test message.");
+            instance.sendMessage("message", config.topics, (testErr) => {
+                if (testErr) {
+                    callback(testErr);
+
+                    return;
+                }
+                console.log("Test message inserted successfully");
+                requeue((reErr) => {
+                    if (reErr) {
+                        callback(reErr);
+
+                        return;
+                    }
+
+                    // Ensure heartbeat cleanup works. If so I will get
+                    // the same message as above after the allotted time
+                    // and I can then remove it.
+                    setTimeout(() => {
+                        cleanupElapsed = true;
+                        console.log("Heartbeat lifetime time elapsed!");
+                    }, instance.config.heartbeatLifetimeSeconds * 1000);
+                    heartbeat((hbErr) => {
```

Comment:
Already discussed but I got confused around this area. Maybe add a comment explaining that the heartbeat function sets up a listener and first "consumes" the message and THEN runs heartbeat on it so that when `remove` is called below and its listener is set up it will not get that message for the duration of the heartbeat.

I think I may have also been confused because this is exactly how you would NOT use it in a real situation. I imagine in a real situation you would set up something like.

```
setTimeout(() => {
    message.heartbeat().
}, 5000); // every five seconds.

// do work to use the message.
// Delete the message and clean up the timeout or something.
```